src/codegen/compile.ml

```diff
@@ -1719,84 +1735,83 @@ module BitTagged = struct
       (* tag *)
-      |> Int32.logor (TaggingScheme.tag_of_typ pty)
+      |> Int64.logor (TaggingScheme.tag_of_typ pty)
     | _ -> assert false)
 
   (* dynamic *)
-  let sanity_check_can_tag_i64 env pty get_x =
+  let sanity_check_can_tag_signed env pty get_x =
     if TaggingScheme.debug || !Flags.sanity then
       get_x ^^
-      Func.share_code2 Func.Always env (prim_fun_name pty "check_can_tag_i64") (("res", I32Type), ("x", I64Type)) [I32Type]
+      Func.share_code2 Func.Always env (prim_fun_name pty "check_can_tag_i64") (("res", I64Type), ("x", I64Type)) [I64Type]
         (fun env get_res get_x -> Type.(
           match pty with
           | Nat | Int | Int64 | Int32 ->
             let sbits = sbits_of pty in
             let lower_bound = Int64.(neg (shift_left 1L sbits)) in
             let upper_bound = Int64.shift_left 1L sbits in
             (* lower_bound <= x < upper_bound *)
-            compile_const_64 lower_bound ^^
+            compile_unboxed_const lower_bound ^^
             get_x ^^
-            G.i (Compare (Wasm.Values.I64 I32Op.LeS)) ^^
-            get_x ^^ compile_const_64 upper_bound ^^
-            G.i (Compare (Wasm.Values.I64 I32Op.LtS)) ^^
-            G.i (Binary (Wasm.Values.I32 I32Op.And))
+            compile_comparison I64Op.LeS ^^
+            get_x ^^ compile_unboxed_const upper_bound ^^
+            compile_comparison I64Op.LtS ^^
+            G.i (Binary (Wasm_exts.Values.I64 I64Op.And))
          | Nat64 | Nat32 ->
             let ubits = ubits_of pty in
             let upper_bound = Int64.shift_left 1L ubits in
             (* 0 <= x < upper_bound *)
-            get_x ^^ compile_const_64 upper_bound ^^
-            G.i (Compare (Wasm.Values.I64 I32Op.LtU))
+            get_x ^^ compile_unboxed_const upper_bound ^^
+            compile_comparison I64Op.LtU
          | _ ->
             assert false) ^^
          get_res ^^
-         G.i (Compare (Wasm.Values.I32 I32Op.Eq)) ^^
+         compile_comparison I64Op.Eq ^^
          E.else_trap_with env (prim_fun_name pty "check_can_tag_i64") ^^
          get_res)
     else
       G.nop
 
-  let if_can_tag_i64 env pty retty is1 is2 = Type.(
+  let if_can_tag_signed env pty retty is1 is2 = Type.(
     match pty with
-    | Nat | Int | Int64 | Int32 ->
+    | Nat | Int | Int64 ->
       Func.share_code1 Func.Never env
-        (prim_fun_name pty "if_can_tag_i64") ("x", I64Type) [I32Type] (fun env get_x ->
+        (prim_fun_name pty "if_can_tag_i64") ("x", I64Type) [I64Type] (fun env get_x ->
         (* checks that all but the low sbits are either all 0 or all 1 *)
         get_x ^^
-        get_x ^^ compile_shrS64_const (Int64.of_int (64 - sbits_of pty)) ^^
-        G.i (Binary (Wasm.Values.I64 I32Op.Xor)) ^^
-        compile_shrU64_const (Int64.of_int (sbits_of pty)) ^^
-        G.i (Test (Wasm.Values.I64 I64Op.Eqz)) ^^
-        sanity_check_can_tag_i64 env pty get_x) ^^
+        get_x ^^ compile_shrS_const (Int64.of_int ((64 - ubits_of pty))) ^^
```

Comment:
Why they change from `sbits_of` to `ubits_of`? Is that deliberate?

Reply:
Thanks! Indeed, this is unintended. I fixed this.
(Luckily, though, if I am not mistaken, it would not have impacted the subsequent logic, as the relevant most significant bits (i.e. `64 - sbits_of pty`) would have been set to the sign bit. This would explain why tests still passed.)